File: 2020-03/3-28/17电话号码的字母组合/17.py
#  -*- coding: utf-8字符串转换整数 -*-

#  Editor      : Pycharm
#  File        : 17
#  Author      : ZhenyuMa
#  Created     : 2020/3/28 11:44 下午
#  Description : 回溯


from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []

        self.record = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8字符串转换整数': 'tuv',
            '9': 'wxyz'
        }
        self.res = []
        self.helper(digits, 0, '')
        return self.res

    def helper(self, digits, idx, cur_digit):
        if idx == len(digits):
            self.res.append(cur_digit)
            return

        for digit in self.record[digits[idx]]:
            self.helper(digits, idx + 1, cur_digit + digit)

File: 2020-04/4-29/17电话号码的字母组合/17.cpp
class Solution {
private:
    unordered_map<char, string> record = {
            {'2', "abc"},
            {'3', "def"},
            {'4', "ghi"},
            {'5', "jkl"},
            {'6', "mno"},
            {'7', "pqrs"},
            {'8', "tuv"},
            {'9', "wxyz"}
        };
        vector<string> res;
public:
    vector<string> letterCombinations(string digits) {
        if(digits.size() == 0)
            return res;

        string cur_res = "";
        helper(digits, 0, cur_res);
        return res;
    }

    void helper(string digits, int start_idx, string& cur_res){
        if(start_idx == digits.size()){
            res.push_back(cur_res);
            return;
        }
        
        for(auto c: record[digits[start_idx]]){
            cur_res += c;
            helper(digits, start_idx + 1, cur_res);
            cur_res.erase(cur_res.size() - 1, 1);
        }
    }
};